### ISSUE TITLE
Partial EPG programs descriptions

### ISSUE THREAD

**Issue by trepmag** [rbuehlma/pvr.zattoo#50]
Hello rbuehlma, many thanks for the add-on; it works great!

I noticed that programs descriptions, for Zattoo provider, is either:
- sometime missing
- when present doesn't corresponds to the one show in the browser version

Tried also with Salt.ch provider and in this case the programs descriptions are always missing...

**Comment by rbuehlma** [comment]
Great to hear that the addon works. Do you have a concrete sample of a mismatching description? Do you use the alternative EPG service? Could you try to clear the EPG cache in the settings and recheck.

**Comment by trepmag** [comment]
I just cleared the guide data and could steel notice program description mismatch.

For example here is a description mismatch on program 'Agitateurs de goût' at 12:50 of channel 'France 5 HD' as follow:
- Zattoo web: "Sur l'étang de Thau, dans l'Hérault, deux générations de la famille Tarbouriech ouvrent les portes de leur sanctuaire dédié à l'huître de Bouzigues."
- Add-on Zattoo provider: "L'huître en Occitanie"

So, descriptions mismatch but both feet the program: they mention both the word 'huître' which means oyster FYI ;-)

Couldn't notice any difference when using the "enhanced EPG service"...

I'm going to inspect log to see if I can find an issue there...

**Comment by rbuehlma** [comment]
Thanks for the additional information. This is a limitation of the Zattoo API and the way Kodi loads the EPG data:
The description of each program needs to be retrieved by an extra request to the Zattoo API. For Kodi, all the EPG data needs to be fetched at once, i.e. its not possible to fetch additional EPG data upon selecting a program. To get this information, it would be required to do thousands of requests.
To solve this issue, I have implemented the extern epg service which does gather all these descriptions and delivers them to Kodi.

You can try to re-enable the external epg service (please note that you enable both options in the addon settings for it) and then again clear the epg cache. You should then have the complete description.

**Comment by trepmag** [comment]
Ok understood that, with Zattoo provider setting both options of the 'External EPG service' work; no description mismatch!!! :-)

Tried the same with Salt.ch provider and the EPG is completely empty; indeed in log I can notice those kind of lines:
```
2019-07-06 13:09:22.923 T:16008   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/france-3/1562324960/1562670560:
```

**Comment by rbuehlma** [comment]
Since I do not have an account for tv.salt.ch I cannot really analyze this case. Could you provide a complete debug log when having salt selected?

**Comment by trepmag** [comment]
Here is it from when the add-on start:
```
2019-07-06 13:09:18.581 T:5516  NOTICE: AddOnLog: Zattoo PVR Client: Using useragent: Kodi/18.3 pvr.zattoo/18.1.3 (Kodi PVR addon)
2019-07-06 13:09:18.845 T:5516  NOTICE: AddOnLog: Zattoo PVR Client: Country code: CH
2019-07-06 13:09:18.845 T:5516  NOTICE: AddOnLog: Zattoo PVR Client: Service region country:
2019-07-06 13:09:18.845 T:5516  NOTICE: AddOnLog: Zattoo PVR Client: Stream type: hls
2019-07-06 13:09:18.845 T:5516  NOTICE: AddOnLog: Zattoo PVR Client: Recall is enabled for 604800 seconds.
2019-07-06 13:09:18.845 T:5516  NOTICE: AddOnLog: Zattoo PVR Client: Selective recall is disabled
2019-07-06 13:09:18.845 T:5516  NOTICE: AddOnLog: Zattoo PVR Client: Recordings are enabled
2019-07-06 13:09:18.899 T:5516  NOTICE: PVR Manager: Starting
2019-07-06 13:09:19.297 T:13452 WARNING: Unable to determine channel type. Defaulting to TV.
2019-07-06 13:09:19.298 T:13452 WARNING: Previous line repeats 3 times.
2019-07-06 13:09:19.298 T:13452  NOTICE: PVR Manager: Started
2019-07-06 13:09:19.303 T:3080  NOTICE: EPG thread started
2019-07-06 13:09:20.576 T:7904   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/hd1/1562324960/1562670560:
2019-07-06 13:09:20.576 T:3148   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/d17/1562324960/1562670560:
2019-07-06 13:09:20.576 T:7904   ERROR: AddOnLog: Zattoo PVR Client: Open URL failed. Try to re-init session.
2019-07-06 13:09:20.577 T:16008   ERROR: Previous line repeats 1 times.
2019-07-06 13:09:20.577 T:16008   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/nt1/1562324960/1562670560:
2019-07-06 13:09:20.577 T:16008   ERROR: AddOnLog: Zattoo PVR Client: Open URL failed. Try to re-init session.
2019-07-06 13:09:20.579 T:12896   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/tf1/1562324960/1562670560:
2019-07-06 13:09:20.579 T:12896   ERROR: AddOnLog: Zattoo PVR Client: Open URL failed. Try to re-init session.
2019-07-06 13:09:20.585 T:6584   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/ab1/1562324960/1562670560:
2019-07-06 13:09:20.586 T:6584   ERROR: AddOnLog: Zattoo PVR Client: Open URL failed. Try to re-init session.
2019-07-06 13:09:20.601 T:3148  NOTICE: AddOnLog: Zattoo PVR Client: Country code: CH
2019-07-06 13:09:20.601 T:3148  NOTICE: AddOnLog: Zattoo PVR Client: Service region country:
2019-07-06 13:09:20.601 T:3148  NOTICE: AddOnLog: Zattoo PVR Client: Stream type: hls
2019-07-06 13:09:20.601 T:3148  NOTICE: AddOnLog: Zattoo PVR Client: Recall is enabled for 604800 seconds.
2019-07-06 13:09:20.601 T:3148  NOTICE: AddOnLog: Zattoo PVR Client: Selective recall is disabled
2019-07-06 13:09:20.601 T:3148  NOTICE: AddOnLog: Zattoo PVR Client: Recordings are enabled
2019-07-06 13:09:20.663 T:3148   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/d17/1562324960/1562670560:
2019-07-06 13:09:20.733 T:3148   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/tmc/1562324960/1562670560:
2019-07-06 13:09:20.733 T:3148   ERROR: AddOnLog: Zattoo PVR Client: Open URL failed. Try to re-init session.
2019-07-06 13:09:20.750 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Country code: CH
2019-07-06 13:09:20.750 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Service region country:
2019-07-06 13:09:20.750 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Stream type: hls
2019-07-06 13:09:20.750 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Recall is enabled for 604800 seconds.
2019-07-06 13:09:20.750 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Selective recall is disabled
2019-07-06 13:09:20.750 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Recordings are enabled
2019-07-06 13:09:20.815 T:16008   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/nt1/1562324960/1562670560:
2019-07-06 13:09:20.889 T:16008   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/6ter/1562324960/1562670560:
2019-07-06 13:09:20.889 T:16008   ERROR: AddOnLog: Zattoo PVR Client: Open URL failed. Try to re-init session.
2019-07-06 13:09:20.918 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Country code: CH
2019-07-06 13:09:20.918 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Service region country:
2019-07-06 13:09:20.918 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Stream type: hls
2019-07-06 13:09:20.918 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Recall is enabled for 604800 seconds.
2019-07-06 13:09:20.918 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Selective recall is disabled
2019-07-06 13:09:20.918 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Recordings are enabled
2019-07-06 13:09:20.983 T:16008   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/6ter/1562324960/1562670560:
2019-07-06 13:09:21.047 T:16008   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/rtl9/1562324960/1562670560:
2019-07-06 13:09:21.047 T:16008   ERROR: AddOnLog: Zattoo PVR Client: Open URL failed. Try to re-init session.
2019-07-06 13:09:21.077 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Country code: CH
2019-07-06 13:09:21.077 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Service region country:
2019-07-06 13:09:21.077 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Stream type: hls
2019-07-06 13:09:21.077 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Recall is enabled for 604800 seconds.
2019-07-06 13:09:21.077 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Selective recall is disabled
2019-07-06 13:09:21.077 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Recordings are enabled
2019-07-06 13:09:21.138 T:16008   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/rtl9/1562324960/1562670560:
2019-07-06 13:09:21.209 T:16008   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/tsr1/1562324960/1562670560:
2019-07-06 13:09:21.209 T:16008   ERROR: AddOnLog: Zattoo PVR Client: Open URL failed. Try to re-init session.
2019-07-06 13:09:21.239 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Country code: CH
2019-07-06 13:09:21.239 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Service region country:
2019-07-06 13:09:21.239 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Stream type: hls
2019-07-06 13:09:21.239 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Recall is enabled for 604800 seconds.
2019-07-06 13:09:21.239 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Selective recall is disabled
2019-07-06 13:09:21.239 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Recordings are enabled
2019-07-06 13:09:21.251 T:12896  NOTICE: AddOnLog: Zattoo PVR Client: Country code: CH
2019-07-06 13:09:21.251 T:12896  NOTICE: AddOnLog: Zattoo PVR Client: Service region country:
2019-07-06 13:09:21.251 T:12896  NOTICE: AddOnLog: Zattoo PVR Client: Stream type: hls
2019-07-06 13:09:21.251 T:12896  NOTICE: AddOnLog: Zattoo PVR Client: Recall is enabled for 604800 seconds.
2019-07-06 13:09:21.251 T:12896  NOTICE: AddOnLog: Zattoo PVR Client: Selective recall is disabled
2019-07-06 13:09:21.251 T:12896  NOTICE: AddOnLog: Zattoo PVR Client: Recordings are enabled
2019-07-06 13:09:21.310 T:16008   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/tsr1/1562324960/1562670560:
2019-07-06 13:09:21.328 T:12896   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/tf1/1562324960/1562670560:
2019-07-06 13:09:21.375 T:16008   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/tsr2/1562324960/1562670560:
2019-07-06 13:09:21.375 T:16008   ERROR: AddOnLog: Zattoo PVR Client: Open URL failed. Try to re-init session.
2019-07-06 13:09:21.395 T:12896   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/canalF_toute_l_histoire/1562324960/1562670560:
2019-07-06 13:09:21.395 T:12896   ERROR: AddOnLog: Zattoo PVR Client: Open URL failed. Try to re-init session.
2019-07-06 13:09:21.402 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Country code: CH
2019-07-06 13:09:21.402 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Service region country:
2019-07-06 13:09:21.402 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Stream type: hls
2019-07-06 13:09:21.402 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Recall is enabled for 604800 seconds.
2019-07-06 13:09:21.402 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Selective recall is disabled
2019-07-06 13:09:21.402 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Recordings are enabled
2019-07-06 13:09:21.421 T:12896  NOTICE: AddOnLog: Zattoo PVR Client: Country code: CH
2019-07-06 13:09:21.421 T:12896  NOTICE: AddOnLog: Zattoo PVR Client: Service region country:
2019-07-06 13:09:21.422 T:12896  NOTICE: AddOnLog: Zattoo PVR Client: Stream type: hls
2019-07-06 13:09:21.422 T:12896  NOTICE: AddOnLog: Zattoo PVR Client: Recall is enabled for 604800 seconds.
2019-07-06 13:09:21.422 T:12896  NOTICE: AddOnLog: Zattoo PVR Client: Selective recall is disabled
2019-07-06 13:09:21.422 T:12896  NOTICE: AddOnLog: Zattoo PVR Client: Recordings are enabled
2019-07-06 13:09:21.465 T:16008   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/tsr2/1562324960/1562670560:
2019-07-06 13:09:21.482 T:12896   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/canalF_toute_l_histoire/1562324960/1562670560:
2019-07-06 13:09:21.528 T:16008   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/mezzo_live/1562324960/1562670560:
2019-07-06 13:09:21.528 T:16008   ERROR: AddOnLog: Zattoo PVR Client: Open URL failed. Try to re-init session.
2019-07-06 13:09:21.546 T:12896   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/latele/1562324960/1562670560:
2019-07-06 13:09:21.546 T:12896   ERROR: AddOnLog: Zattoo PVR Client: Open URL failed. Try to re-init session.
2019-07-06 13:09:21.560 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Country code: CH
2019-07-06 13:09:21.560 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Service region country:
2019-07-06 13:09:21.560 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Stream type: hls
2019-07-06 13:09:21.560 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Recall is enabled for 604800 seconds.
2019-07-06 13:09:21.560 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Selective recall is disabled
2019-07-06 13:09:21.560 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Recordings are enabled
2019-07-06 13:09:21.627 T:16008   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/mezzo_live/1562324960/1562670560:
2019-07-06 13:09:21.692 T:16008   ERROR: CCurlFile::Open failed with code 404 for https://zattoo.buehlmann.net/epg/api/Epg//667860d0319b8f0ad04f3b3919af47ab/i_tele/1562324960/1562670560:
2019-07-06 13:09:21.692 T:16008   ERROR: AddOnLog: Zattoo PVR Client: Open URL failed. Try to re-init session.
2019-07-06 13:09:21.726 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Country code: CH
2019-07-06 13:09:21.726 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Service region country:
2019-07-06 13:09:21.726 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Stream type: hls
2019-07-06 13:09:21.726 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Recall is enabled for 604800 seconds.
2019-07-06 13:09:21.726 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Selective recall is disabled
2019-07-06 13:09:21.726 T:16008  NOTICE: AddOnLog: Zattoo PVR Client: Recordings are enabled
2019-07-06 13:09:21.754 T:7904  NOTICE: AddOnLog: Zattoo PVR Client: Country code: CH
2019-07-06 13:09:21.754 T:7904  NOTICE: AddOnLog: Zattoo PVR Client: Service region country:
2019-07-06 13:09:21.754 T:7904  NOTICE: AddOnLog: Zattoo PVR Client: Stream type: hls
2019-07-06 13:09:21.754 T:7904  NOTICE: AddOnLog: Zattoo PVR Client: Recall is enabled for 604800 seconds.
2019-07-06 13:09:21.754 T:7904  NOTICE: AddOnLog: Zattoo PVR Client: Selective recall is disabled
2019-07-06 13:09:21.754 T:7904  NOTICE: AddOnLog: Zattoo PVR Client: Recordings are enabled
...
```

**Comment by rbuehlma** [comment]
Ok, one problem is the missing service region:
`2019-07-06 13:09:20.601 T:3148  NOTICE: AddOnLog: Zattoo PVR Client: Service region country:`
I will do a workaround for that.

Could you add an additional log having the external epg service deactivated. Would be interesting to see why the epg is missing there as well.

**Comment by trepmag** [comment]
I tried without the external epg service option for Salt.tv provder and the EPG is present with the "partial" program description; like for the Zattoo provider... It's my  fault when I reported initially it was always missing; I probably missed to clear the guide data...

**Comment by rbuehlma** [comment]
Thats good news. I'll fix the external epg service for salt.

**Comment by rbuehlma** [comment]
Fixed in 18.1.4

**Comment by trepmag** [comment]
Hello rbuehlma, very nice; full description is now shown...

PS: this might produced a regression with recording trigger failing but will try again with previous version; will post a separate issue if appropriate...
